nx/lib/nx/lin_alg.ex

```diff
@@ -554,23 +555,23 @@ defmodule Nx.LinAlg do
   ### Error cases
 
       iex> Nx.LinAlg.invert(Nx.tensor([[3, 0, 0, 0], [2, 1, 0, 0]]))
-      ** (ArgumentError) can only invert square matrices, got: {2, 4}
+      ** (ArgumentError) expected tensor to match shape {n, n}, got tensor with shape {2, 4}
 
       iex> Nx.LinAlg.invert(Nx.tensor([[0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [1, 1, 1, 1]]))
       ** (ArgumentError) can't solve for singular matrix
 
   """
   @doc from_backend: false
-  def invert(tensor) do
-    %T{shape: {m, n}} = tensor = Nx.to_tensor(tensor)
-
-    if m != n do
-      raise ArgumentError,
-            "can only invert square matrices, got: {#{m}, #{n}}"
-    end
+  defn invert(tensor) do
+    assert_shape_pattern(tensor, {n, n})
 
     identity = Nx.eye(tensor)
-    solve(tensor, identity)
+    inverted = Nx.LinAlg.solve(tensor, identity)
+
+    Nx.Defn.Kernel.custom_grad(inverted, fn ans, g ->
+      ans_t = Nx.transpose(ans)
+      ans_t |> Nx.negate() |> Nx.dot(g) |> Nx.dot(ans_t)
```

Comment:
@josevalim I couldn't figure out how to use `-1 |> Nx.multiply(ans_t)` here. In this case `Nx.negate` did the job fine (still needs tests, though), but I'd like to know what's the proper way to do things.

This still in draft because I need to validate the grad

Reply:
That should have worked. Do you get an error or something?

Reply:
Yes. This is the error:

```elixir
== Compilation error in file lib/nx/lin_alg.ex ==
** (CompileError) lib/nx/lin_alg.ex:572: invalid numerical expression:

    -1

    lib/nx/defn/compiler.ex:529: Nx.Defn.Compiler.compile_error!/3
    (elixir 1.13.0) lib/enum.ex:1715: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
    lib/nx/defn/compiler.ex:379: Nx.Defn.Compiler.normalize/2
    (elixir 1.13.0) lib/enum.ex:1715: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
    lib/nx/defn/compiler.ex:379: Nx.Defn.Compiler.normalize/2
    (elixir 1.13.0) lib/enum.ex:1715: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
    lib/nx/defn/compiler.ex:379: Nx.Defn.Compiler.normalize/2
    (elixir 1.13.0) lib/enum.ex:1715: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
```

For the snippet `-1 |> Nx.multiply(ans_t)` instead of `ans_t |> Nx.negate()`

Reply:
@polvalente perhaps we didn't implement unary minus as part of Nx.Defn.Kernel? o.O

Reply:
Oh, it is happening because of this at the top `import Nx.Defn.Kernel, only: [keyword!: 2]` of LinAlg. Change it to `require Nx.Defn.Kernel`, replace all `keyword!` calls by `Nx.Defn.Kernel.keyword!` and you should be good!

Reply:
Actually, I got another fix, hold on.

Reply:
Fixed on main!